### PR TITLE
feat(completion): support no space format of short flags

### PR DESCRIPTION
**Pull request by fiftin**
Cobra supports a "no space" format for short flags but doesn't support auto complete for this case.
 
Related to issue https://github.com/spf13/cobra/issues/1629

### CODE REVIEW

**Comment by marckhouzam** [comment]
Supet exciting @fiftin !  I didn't expect it to be that simple 😁.  I'll give it a try ASAP 🚀

You will have to sign the CLA before we can accept the contribution however.

**Comment by fiftin** [comment]
@marckhouzam ,

How to do this?

**Comment by marckhouzam** [comment]
@fiftin I'm having trouble understanding what this change is supposed to provide.
Could you provide an example output of before and after the change please?

**Comment by fiftin** [comment]
-f123<TAB>

Output

Before: Empty

After:

-f flag description

**Comment by marckhouzam** [comment]
@fiftin Sorry but I'm still having trouble understanding the behaviour you seek.

If a program has a short flag `-f` like your example, and the user does:
`prog -f123<TAB>` why would the completion show `-f`?  And what is the `123` string; is it a value for `-f` or are they other short flags?

**Comment by marckhouzam** [comment]
@fiftin 
So let's say the user currently does:
```
$ kubectl -n infra-pro<TAB>
# or
$ kubectl -n=infra-pro<TAB>
```
in both cases, the completion logic would suggest all namespaces that start with `infra-pro`, for example `infra-production1` and `infra-production2`.

I believe the same behaviour should happen when there is no space after the short flag.
```
$ kubectl -ninfra-pro<TAB>
infra-production1      infra-production2
```

But there are special cases we would need to treat differently:
```
$ kubectl -n<TAB>
# Should probably complete to -n

$ kubectl -itn<TAB>
# Should probably complete to -n because each letter 'i', 't', 'n' represent a valid short flag and the first two don't take an argument

$ kubectl -itni<TAB>
# Should probably complete to all namespaces starting with 'i' because the first letter 'i' and second letter 't', represent
# valid short flags not needing an argument, but the third letter 'n' represents a short flag that requires a namespace
# as an argument, so the forth letter 'i' no longer represents a flag but the prefix of a namespace name.
```

Does this make sense to you?

**Comment by fiftin** [comment]
Hi @marckhouzam,

Done. Please see test `TestShorthandFlagCompletionInGoWithDesc` for details.

https://github.com/spf13/cobra/blob/87c16f9d0b2dccac2ca8911441cfe21f1ee52d17/completions_test.go#L476

**Review by marckhouzam — CHANGES_REQUESTED** [review]
This is a great start @fiftin, thanks!

Testing with a real program, the go logic to handle short flags without a space seems to work as expected.

I like your idea of checking if a completion function exists to decide if
we should complete a short flag value, or if we should add a space after
the short flag.

So, when testing with the __complete command things seem good nicely.

However when testing with the actual shell while performing completions
things don't work just yet.  For zsh for example, we need to set the variable `flagPrefix`
in the script.  What I mean is that if we do
```
$ helm -n<tab>
```
and get the completion choices 'firstns' and 'secondns', the zsh completion
script must convert them into '-nfirstns' and '-nsecondns'.  We already
do this for the = case:
```
$ helm -n=<tab>
firstns       secondns
```
So the zsh shell script would need to be modified a little to handle the new case
of completion of short flags.  The completion for bash v2 will also need
to be fixed (I haven't looked at the details, but it also does not work).  And we'll need to
try things out for fish and powershell to see what they need.  I can help
out with all this if you want, you just let me know.

But we are making good progress 👍

File: completions_test.go [hunk 1]
```diff
@@ -2280,7 +2350,7 @@ func removeCompCmd(rootCmd *Command) {
 	}
 }
 
-func TestDefaultCompletionCmd(t *testing.T) {
+func xTestDefaultCompletionCmd(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (fiftin) [reply completions_test.go]: Hi @marckhouzam I did. I updated PR yesterday with --force flag

**Comment by fiftin** [comment]
Hi @marckhouzam , any updates? What do you think about our issue?

**Comment by marckhouzam** [comment]
Thanks @fiftin for the effort.  I will need to take the time to give this proper attention but I will not have time for the next couple of weeks.  I'll get back to this as soon as I can.  Thanks for your patience

**Comment by fiftin** [comment]
Hi @marckhouzam ,

Do you have a time to work on the issue?

**Comment by marckhouzam** [comment]
@fiftin Are you no longer interested in making this contribution?  I was about to review it when I saw you had closed it.
I'm sorry it took a long time, but it is summer time and I was on vacation.

**Pull request closed without merging**